src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySourceConfig.java

```diff
@@ -133,14 +133,23 @@ public String getDatasetProject() {
   public void validate(FailureCollector collector) {
     super.validate(collector);
     String bucket = getBucket();
-    if (!containsMacro("bucket") && bucket != null) {
-      // Basic validation for allowed characters as per https://cloud.google.com/storage/docs/naming
-      Pattern p = Pattern.compile("[a-z0-9._-]+");
-      if (!p.matcher(bucket).matches()) {
-        collector.addFailure("Bucket contains characters other than lowercase characters, numbers, '.', '_', and '-'.",
-                             null)
-          .withConfigProperty(NAME_BUCKET);
-      }
+
+    if (!containsMacro(bucket)) {
+      // Allowed character validation for bucket name as per https://cloud.google.com/storage/docs/naming
+      String errorMessage = "Bucket name can only contain lowercase letters, numbers, '.', '_', and '-'.";
+      BigQueryUtil.match(bucket, NAME_BUCKET, BigQueryUtil.BUCKET_PATTERN, collector, errorMessage);
+    }
+
+    if (!containsMacro(dataset)) {
+      // Allowed character validation for dataset name as per https://cloud.google.com/bigquery/docs/datasets
+      String errorMessage = "Dataset name can only contain letters (lower or uppercase), numbers and '_'.";
+      BigQueryUtil.match(dataset, NAME_DATASET, BigQueryUtil.DATASET_PATTERN, collector, errorMessage);
+    }
+
+    if (!containsMacro(table)) {
+      // Allowed character validation for table name as per https://cloud.google.com/bigquery/docs/tables
+      String errorMessage = "Table name can only contain letters (lower or uppercase), numbers and '_'.";
```

Comment:
actually, since there multiple places this is called, I think it makes sense to have separate validateBucket(), validateDataset(), validateTable() methods and also move the error messages into the util class as constants.